File: src/CalendarPrinter/Logic/MonthlyCalendarGenerator.cs
using CalendarPrinter.Model;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace CalendarPrinter.Logic
{
    internal abstract class MonthlyCalendarGenerator : CalendarGenerator
    {

        protected IEnumerable<DayOfWeek> BuildWeekdayOrder()
        {
            yield return DayOfWeek.Monday;
            yield return DayOfWeek.Tuesday;
            yield return DayOfWeek.Wednesday;
            yield return DayOfWeek.Thursday;
            yield return DayOfWeek.Friday;
            yield return DayOfWeek.Saturday;
            yield return DayOfWeek.Sunday;
        }

        protected IEnumerable<List<CalendarCell>> CalculateCalendarRows(IEnumerable<DateTime> dates, EventCalendar eventCalendar)
        {
            var datesEnumerator = CreateCells(dates, eventCalendar).GetEnumerator();
            if (!datesEnumerator.MoveNext())
                throw new InvalidOperationException();

            var weekdays = BuildWeekdayOrder().ToArray();

            // First row
            var row = new List<CalendarCell>(weekdays.Length);

            foreach (var item in weekdays)
            {
                if (datesEnumerator.Current.Date.Value.DayOfWeek != item)
                    row.Add(CalendarCell.Empty);
                else
                {
                    row.Add(datesEnumerator.Current);

                    if (!datesEnumerator.MoveNext())
                        break;
                }
            }

            Debug.Assert(row.Count == weekdays.Length);
            yield return row;
            row = null;

            // Middle rows
            while (true)
            {
                if (datesEnumerator.Current.Date.Value.DayOfWeek == weekdays.First())
                {
                    if (row != null)
                        throw new InvalidOperationException();

                    row = new List<CalendarCell>(weekdays.Length);
                }

                row.Add(datesEnumerator.Current);

                if (datesEnumerator.Current.Date.Value.DayOfWeek == weekdays.Last())
                {
                    Debug.Assert(row.Count == weekdays.Length);
                    yield return row;
                    row = null;
                }

                if (!datesEnumerator.MoveNext())
                    break;
            }


            // Last row
            if (datesEnumerator.Current.Date.Value.DayOfWeek != weekdays.Last())
            {
                bool previousDaysWritten = false;

                foreach (var item in weekdays)
                {
                    if (previousDaysWritten)
                    {
                        row.Add(CalendarCell.Empty);
                    }
                    else
                    {
                        previousDaysWritten = datesEnumerator.Current.Date.Value.DayOfWeek == item;
                    }
                }

                Debug.Assert(row.Count == weekdays.Length);
                yield return row;
            }

        }
    }
}

File: src/CalendarPrinter/MonthlyCalendarGenerator.py
from typing import Iterator
from .CalendarGenerator import CalendarGenerator, CalendarCell
from .DayOfWeek import DayOfWeek
from .generator_with_current import generator_with_current
from .EventCalendar import EventCalendar
   
class MonthlyCalendarGenerator(CalendarGenerator):
    def __init__(self):
        pass

    def BuildWeekdayOrder() -> Iterator[DayOfWeek]:
        yield DayOfWeek.Monday
        yield DayOfWeek.Tuesday
        yield DayOfWeek.Wednesday
        yield DayOfWeek.Thursday
        yield DayOfWeek.Friday
        yield DayOfWeek.Saturday
        yield DayOfWeek.Sunday

    def CalculateCalendarRows(self, dates, eventCalendar: EventCalendar) -> Iterator[list[CalendarCell]]:
        datesEnumerator = generator_with_current(self.CreateCells(dates, eventCalendar))
        if next(datesEnumerator) is None:
            raise GeneratorExit()

        weekdays = list(MonthlyCalendarGenerator.BuildWeekdayOrder())

        # First row
        row = []

        for item in weekdays:
            if datesEnumerator.current.date.isoweekday() != int(item):
                row.append(CalendarCell.Empty)
            else:
                row.append(datesEnumerator.current)

                if next(datesEnumerator) is None:
                    break
                    
        assert len(row) == len(weekdays)
        yield row
        row = None

        # Middle rows
        while True:
            if datesEnumerator.current.date.isoweekday() == int(weekdays[0]):
                if row is not None:
                    raise RuntimeError()

                row = []

            row.append(datesEnumerator.current)

            if datesEnumerator.current.date.isoweekday() == int(weekdays[-1]):
                assert len(row) == len(weekdays)
                yield row
                row = None

            if next(datesEnumerator) is None:
                break


        # Last row
        if datesEnumerator.current.date.isoweekday() != int(weekdays[-1]):
            previousDaysWritten = False

            for item in weekdays:
                if previousDaysWritten:
                    row.append(CalendarCell.Empty)
                else:
                    previousDaysWritten = (datesEnumerator.current.date.isoweekday() == int(item))

            assert len(row) == len(weekdays)
            yield row
